Ndev 2981 trace call overrides params

romanova-natasha opened the pull request:
1) `eth_call` with overriding is not supported yet (task for it NDEV-2400)
2) for `blockOverrides` we support only `time` and `number` parameters

romanova-natasha committed:
add case to check stateOverrides nonce

add case to check stateOverrides nonce

romanova-natasha committed:
add dbg prints and run only one test

add dbg prints and run only one test

romanova-natasha committed:
Merge branch 'develop' into NDEV-2981-traceCall-overrides-param

romanova-natasha committed:
mv tests to new file, add cases for stateOverride check

mv tests to new file, add cases for stateOverride check

romanova-natasha committed:
add cases for stateOverride check

add cases for stateOverride check

romanova-natasha committed:
rm dbg

rm dbg

romanova-natasha committed:
link bugs to tests and skip these cases

link bugs to tests and skip these cases

romanova-natasha committed:
add blockOverride case using debug_traceCall request

add blockOverride case using debug_traceCall request

romanova-natasha committed:
Merge branch 'run-specific-test' into NDEV-2981-traceCall-overrides-param

romanova-natasha committed:
Merge branch 'run-specific-test' into NDEV-2981-traceCall-overrides-param

romanova-natasha committed:
add case with blockOverrides for debug_traceCall method

add case with blockOverrides for debug_traceCall method

romanova-natasha committed:
add comment to test_stateOverrides_eth_call_override_code case

add comment to test_stateOverrides_eth_call_override_code case

romanova-natasha committed:
minor refactoring

minor refactoring

romanova-natasha committed:
add negative cases to override params suite, minor refactoring, add comment to cases with useless error message, skip one case due to bug 3010

add negative cases to override params suite, minor refactoring, add comment to cases with useless error message, skip one case due to bug 3010

romanova-natasha committed:
add cases for blockOverrides with time parameter

add cases for blockOverrides with time parameter

romanova-natasha committed:
add negative cases for stateDiff param

add negative cases for stateDiff param

romanova-natasha committed:
add cases for stateOverride nonce, balance

add cases for stateOverride nonce, balance

romanova-natasha committed:
add cases for stateOverride code parameter

add cases for stateOverride code parameter

romanova-natasha committed:
enable stateDiff cases which were fixed

enable stateDiff cases which were fixed

romanova-natasha committed:
fix stateOverride balance tests, add a negative case: insufficient balance for executing tx

fix stateOverride balance tests, add a negative case: insufficient balance for executing tx

romanova-natasha committed:
add blockOverrides time and number case, minor refactoring, add case for stateOverrides lower nonce

add blockOverrides time and number case, minor refactoring, add case for stateOverrides lower nonce

romanova-natasha committed:
add case: stateOverrides and blockOverrides in one requests, add case stateOverrides all params without storage, small refactoring

add case: stateOverrides and blockOverrides in one requests, add case stateOverrides all params without storage, small refactoring

romanova-natasha committed:
refactoring of debug_methods suite according to the recent changes

refactoring of debug_methods suite according to the recent changes

romanova-natasha committed:
add stateOverride cases for state param

add stateOverride cases for state param

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+    storage_value = 57
+
+    @pytest.fixture(scope="class")
+    def retrieve_block_tx(self, storage_contract, web3_client):
```
you have self.web3_client here

romanova-natasha replied:
yep, agree) already fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+        sender_account = self.accounts[0]
+        nonce = web3_client.eth.get_transaction_count(sender_account.address)
+        instruction_tx = storage_contract.functions.storeBlock().build_transaction(
+            {
```
why didn't you use web3_client.make_raw_tx for creating this object?

romanova-natasha replied:
just forget about it, fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+
+        address_from = call_storage_tx["from"].lower()
+        override_params = {"stateOverrides": {address_from: {"nonce": 17}}, "tracer": "prestateTracer"}
+        params.pop(2)
```
what did you delete here and why?

romanova-natasha replied:
We have different tracers, I use `prestateTracer` to check that every overrided param is actually was overrided before emulation. I delete this object `{"tracer": "prestateTracer"}` sometimes cause I want ether to rewrite params or call `tracer` instead of `prestateTracer`

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+        address_to = call_storage_tx["to"].lower()
+        params = self.fill_params_for_storage_contract_trace_call(call_storage_tx)
+
+        params.append({"tracer": "prestateTracer"})
```
it can be moved to fill_params_for_storage_contract_trace_call as it is repeated every time

romanova-natasha replied:
We don't need to call `prestateTracer` everywhere, sometimes we need just `trcaer`

romanova-natasha replied:
i will do 2 methods fill_params_for_storage_contract_trace_call and fill_params_for_storage_contract_trace_call_prestate, it will reduce visual noice of params.append({"tracer": "prestateTracer"})

romanova-natasha replied:
done

kristinaNikolaevaa replied:
it can be 1 function with parameter is_prestateTracer or something and default value

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+        assert "error" not in response_overrided, "Error in response"
+        #code
+        assert response_overrided["result"][address_to]["code"] != response["result"][address_to]["code"]
+        assert response_overrided["result"][address_to]["code"] == "0x6080604052348015"
```
Lets add tests with overriding [address_to]["code"] to check that it affect on result of call

romanova-natasha replied:
added, `tests test_stateOverrides_debug_traceCall_override_code_*`

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+from utils.accounts import EthAccounts
+from utils.tracer_client import TracerClient
+from utils.helpers import padhex
+from integration.tests.tracer.test_tracer_historical_methods import call_storage, store_value
```
let's not import test files in tests files 
move it in some common files please

romanova-natasha replied:
done.
moved everything we have for storage contract to a separate class

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+    def call_store_value_tx(self, storage_contract, web3_client):
+        sender_account = self.accounts[0]
+        receipt = store_value(sender_account, self.storage_value, storage_contract, web3_client)
+        return self.web3_client.wait_get_transaction_by_hash(receipt["transactionHash"].hex())
```
the store_value function include waiting for transaction, you don't need it
looks like fixture call_store_value_tx complitely the same as store_value function

romanova-natasha replied:
`store_value` returns receipt of tx, `call_store_value_tx` returns tx object

kristinaNikolaevaa commented on integration/tests/tracer/test_override_params.py:
```diff
@@ -0,0 +1,800 @@
+    def call_storage_tx(self, storage_contract, web3_client):
+        sender_account = self.accounts[0]
+        _, _, receipt = call_storage(sender_account, storage_contract, self.storage_value, "blockNumber", web3_client)
+        return self.web3_client.wait_get_transaction_by_hash(receipt["transactionHash"].hex())
```
the same comment as about call_store_value_tx

romanova-natasha replied:
I'm getting transaction by hash cause I need to fill params for debug_traceTransaction, there is not enough info in tx receipt

kristinaNikolaevaa replied:
okey, I got it
but you still don't need to wait it, becauce call_storage include it

romanova-natasha replied:
fixed

romanova-natasha committed:
fix tests stateOverrides with code param: add check of trace with overrided code for contract and externally-owned accounts

fix tests stateOverrides with code param: add check of trace with overrided code for contract and externally-owned accounts

romanova-natasha committed:
add contract function with retrieving sender balance, check balance is overrided

add contract function with retrieving sender balance, check balance is overrided

romanova-natasha committed:
fix: add make_raw_tx where it is nedeed

fix: add make_raw_tx where it is nedeed

romanova-natasha committed:
add function: fill params for prestateTracer call

add function: fill params for prestateTracer call

romanova-natasha committed:
mv storage contract to separate class

mv storage contract to separate class

romanova-natasha committed:
fix review

fix review

romanova-natasha committed:
fix review: do not wait transaction

fix review: do not wait transaction

romanova-natasha committed:
add case for stateDiff: check storage with 2 indixes

add case for stateDiff: check storage with 2 indixes

romanova-natasha committed:
rm code field from state and stateDiff requests

rm code field from state and stateDiff requests

romanova-natasha committed:
fix state/stateDiff param format: should be hex string 32 bytes

fix state/stateDiff param format: should be hex string 32 bytes

romanova-natasha committed:
fix state/stateDiff param format: should be hex string 32 bytes

fix state/stateDiff param format: should be hex string 32 bytes

romanova-natasha committed:
fix blockOverrides: add is_tx_block flag to control block number in requests

fix blockOverrides: add is_tx_block flag to control block number in requests

romanova-natasha committed:
fix stateDiff cases

fix stateDiff cases

romanova-natasha committed:
rm dbg info

rm dbg info

romanova-natasha committed:
fix state cases

fix state cases

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-2981-traceCall-overrides-param

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_methods.py:
```diff
@@ -37,43 +39,33 @@ def validate_response_result(self, response):
-            timeout_sec=10,
-        )
-        tx_info = self.web3_client.get_transaction_by_hash(tx_hash)
+        tx_info = self.web3_client.wait_get_transaction_by_hash(receipt["transactionHash"].hex())
```
send_neon contains waiting for a transaction

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_methods.py:
```diff
@@ -93,17 +85,12 @@ def test_debug_trace_call_zero_eth_call(self):
-        tx_info = self.web3_client.get_transaction_by_hash(tx_hash)
+        _, _, receipt = storage_object.call_storage(sender_account, store_value, "blockNumber")
+       
+        tx_info = self.web3_client.wait_get_transaction_by_hash(receipt["transactionHash"].hex())
```
call_storage contains waiting

romanova-natasha replied:
fixed

romanova-natasha committed:
rm wait for get transaction method

rm wait for get transaction method

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-2981-traceCall-overrides-param

# Conflicts:
#	integration/tests/tracer/test_tracer_debug_methods.py

kristinaNikolaevaa commented on utils/web3client.py:
```diff
@@ -109,6 +110,17 @@ def get_transaction_by_hash(self, transaction_hash):
             return None
+    
+    @allure.step("Get transaction by hash with wait for result is available")
+    def wait_get_transaction_by_hash(self, transaction_hash, timeout=10):
```
it can be deleted now

romanova-natasha committed:
rm wait transaction method

rm wait transaction method

kristinaNikolaevaa reviewed (changes requested):
CAN BE MERGED ONLY WITH https://github.com/neonlabsorg/neon-evm/pull/425
